stream/src/main/scala/org/apache/pekko/stream/impl/TraversalBuilder.scala

```diff
@@ -380,12 +380,53 @@ import pekko.util.unused
     }
   }
 
+  /**
+   * Try to find `SingleSource` or wrapped such. This is used as a
+   * performance optimization in FlattenConcat and possibly other places.
+   * @since 1.2.0
+   */
+  @InternalApi def getValuePresentedSource[A >: Null](
+      graph: Graph[SourceShape[A], _]): OptionVal[Graph[SourceShape[A], _]] = {
+    def isValuePresentedSource(graph: Graph[SourceShape[_ <: A], _]): Boolean = graph match {
+      case _: SingleSource[_] | _: FutureSource[_] | _: IterableSource[_] | _: JavaStreamSource[_, _] |
+          _: FailedSource[_] =>
+        true
+      case maybeEmpty if isEmptySource(maybeEmpty) => true
+      case _                                       => false
+    }
+    graph match {
+      case _ if isValuePresentedSource(graph) => OptionVal.Some(graph)
+      case _ =>
+        graph.traversalBuilder match {
+          case l: LinearTraversalBuilder =>
+            l.pendingBuilder match {
+              case OptionVal.Some(a: AtomicTraversalBuilder) =>
+                a.module match {
+                  case m: GraphStageModule[_, _] =>
+                    m.stage match {
+                      case _ if isValuePresentedSource(m.stage.asInstanceOf[Graph[SourceShape[A], _]]) =>
+                        // It would be != EmptyTraversal if mapMaterializedValue was used and then we can't optimize.
+                        if ((l.traversalSoFar eq EmptyTraversal) && !l.attributes.isAsync)
+                          OptionVal.Some(m.stage.asInstanceOf[Graph[SourceShape[A], _]])
+                        else OptionVal.None
+                      case _ => OptionVal.None
+                    }
+                  case _ => OptionVal.None
+                }
+              case _ => OptionVal.None
+            }
+          case _ => OptionVal.None
+        }
+    }
+  }
+
   /**
    * Test if a Graph is an empty Source.
    */
   def isEmptySource(graph: Graph[SourceShape[_], _]): Boolean = graph match {
     case source: scaladsl.Source[_, _] if source eq scaladsl.Source.empty => true
     case source: javadsl.Source[_, _] if source eq javadsl.Source.empty() => true
+    case EmptySource                                                      => true
```

Comment:
this doesn't look related to this change - is this a bug that needs to be fixed in its own PR?

Reply:
It's not a bug, just a missing optimization case.

Reply:
I thought that the cases were evaluated in order so this would not change performance in that case and it would actually lead to a different result if EmptySource was matched but that neither of the first 2 cases matched.

Reply:
these methods are internal methods acutually.

Reply:
I still don't understand why this can't be in its own PR. If this is an issue, it looks like it should be backported to 1.1.x.

Reply:
You said no performance optimization should be backported.

Reply:
I can extract it in a separated pr, I will back my hometown tomorrow

Reply:
this is not a perf change - it is a behaviour change

Before this change, `case EmptySource` would not be there and then EmptySource would lead to a return value of false.

Reply:
But this method is only been used for avoid optimization

Reply:
I am lost. The only way this is an optimisation is if EmptySource appears before one or more the existing cases.

Reply:
Should  I  extract this, I just turn on my computer, Tomorrow this time I will at the airport:)